GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs

```diff
@@ -333,12 +366,115 @@ public void Clear()
+        }
+
+        public void EnsureRowVisible(int row)
+        {
+            int countVisible = DisplayedRowCount(includePartialRow: false);
+            int firstVisible = FirstDisplayedScrollingRowIndex;
+            if (row >= 0 && (row < firstVisible || firstVisible + countVisible <= row))
+            {
+                FirstDisplayedScrollingRowIndex = row;
+            }
+        }
+
+        public ObjectId? GetFirstNotSelectedObjectId()
+        {
+            if (_loadedToBeSelectedRevisionsCount > 0 || ToBeSelectedObjectIds.Count == 0)
+            {
+                // At least one of the revisions was selected
+                return null;
+            }
+
+            return ToBeSelectedObjectIds[0];
+        }
+
+        public void ToBeSelectFirstFoundParent(IEnumerable<ObjectId> headParents)
+        {
+            if (headParents is not null)
+            {
+                foreach (var parentId in headParents)
+                {
+                    if (_revisionGraph.TryGetRowIndex(parentId, out int _))
+                    {
+                        ToBeSelectedObjectIds = new ObjectId[] { parentId };
+                        _loadedToBeSelectedRevisionsCount = ToBeSelectedObjectIds.Count;
+                        break;
+                    }
+                }
+            }
+        }
+
         public void LoadingCompleted()
         {
+            if (_loadedToBeSelectedRevisionsCount < ToBeSelectedObjectIds.Count)
+            {
+                // All expected revisions not found, settle with partial (empty) match
+                _loadedToBeSelectedRevisionsCount = ToBeSelectedObjectIds.Count;
+            }
+
+            if (_loadedToBeSelectedRevisionsCount > 0 && _revisionGraph.Count > 0)
+            {
+                // Rows have not been selected yet
+                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    int scrollTo = GetGraphIndexes().Count > 0 ? GetGraphIndexes().Max() : 0;
+                    await this.SwitchToMainThreadAsync();
+                    if (RowCount - 1 < scrollTo)
+                    {
+                        // Wait for the background thread to load all rows in the grid
+                        while (RowCount - 1 < scrollTo)
+                        {
+                            int maxScroll = Math.Min(RowCount - 1, scrollTo);
+                            EnsureRowVisible(maxScroll);
+
+                            UpdateVisibleRowRange();
+                            await Task.Delay(25);
+                        }
+                    }
+                    else
+                    {
+                        // Rows already selected once, reselect and refresh
+                        SelectRowsIfReady(RowCount);
+
+                        if (GetGraphIndexes().Count > 0)
+                        {
+                            EnsureRowVisible(GetGraphIndexes()[0]);
+                        }
```

Comment:
This is not good, we're calling `GetGraphIndexes()` several times. Please refactor to invoke the method and assign the result to a local variable.

Reply:
GetGraphIndexes() is doing the caching.
The information may be cleared as well so info may not be instaciated on more than line 426 and 445.
This could have been set as a Lazy variable too, but as there is no method to Clear a Lazy variable, the init would have to be done in several locations.

Reply:
This still makes me very uneasy. One of the aspects that worries me is threading - the field can be accessed by multiple threads, and could lead to NRE. In c427e589ceefe978927e215c037d0ab3852ecbab I proposed to make the field non-nullable, so we don't risk NREs, though we still have a risk of a race condition. By looking at the other existing fields/properties in use - the risk is probably not significant (though still real).

Reply:
What exactly?
DataGrid is not thread safe, but the same thread basically updates the data for a certain object, I did not find any potential race conditions.